std-bits/table/src/main/java/org/enso/table/write/ExcelWriter.java

```diff
@@ -0,0 +1,300 @@
+        row.createCell(i + firstColumn - 1, CellType.STRING).setCellValue(columns[i].getName());
+      }
+      currentRow++;
+    }
+
+    if (rowCount == 0) {
+      return;
+    }
+
+    Storage[] storages = Arrays.stream(columns).map(Column::getStorage).toArray(Storage[]::new);
+    for (int i = 0; i < rowCount; i++) {
+      Row row = sheet.getRow(currentRow);
+      if (row == null) {
+        row = sheet.createRow(currentRow);
+      }
+
+      for (int j = 0; j < columns.length; j++) {
+        Storage storage = storages[j];
+
+        Cell cell = row.getCell(j + firstColumn - 1);
+        if (cell == null) {
+          cell = row.createCell(j + firstColumn - 1);
+        }
+
+        writeValueToCell(cell, i, storage, workbook);
+      }
+      currentRow++;
+    }
+
+    workbook.setForceFormulaRecalculation(true);
+  }
+
+  private static CellStyle getDateTimeStyle(Workbook workbook, String format) {
+    for(int i = 0; i < workbook.getNumCellStyles(); i++) {
+      CellStyle style = workbook.getCellStyleAt(i);
+      if (style.getDataFormatString().equals(format)) {
+        return style;
+      }
+    }
+
+    CellStyle newStyle = workbook.createCellStyle();
+    newStyle.setDataFormat(workbook.createDataFormat().getFormat(format));
+    return newStyle;
+  }
+
+  private static void writeValueToCell(Cell cell, int j, Storage storage, Workbook workbook) {
+    if (storage.isNa(j)) {
+      cell.setBlank();
+    } else if (storage instanceof DoubleStorage doubleStorage) {
+      cell.setCellValue(doubleStorage.getItem(j));
+    } else if (storage instanceof LongStorage longStorage) {
+      cell.setCellValue(longStorage.getItem(j));
+    } else if (storage instanceof BoolStorage boolStorage) {
+      cell.setCellValue(boolStorage.getItem(j));
+    } else {
+      Object value = storage.getItemBoxed(j);
+      switch (value) {
+        case String s -> cell.setCellValue(s);
+        case Boolean b -> cell.setCellValue(b);
+        case Double d -> cell.setCellValue(d);
+        case Long l -> cell.setCellValue(l);
+        case LocalDateTime ldt -> {
+          cell.setCellValue(ldt);
+          cell.setCellStyle(getDateTimeStyle(workbook, "yyyy-MM-dd HH:mm:ss"));
+        }
+        case LocalDate ld -> {
+          cell.setCellValue(ld);
+          cell.setCellStyle(getDateTimeStyle(workbook, "yyyy-MM-dd"));
+        }
+        case LocalTime lt -> {
+          cell.setCellValue(lt.toSecondOfDay() / SECONDS_IN_A_DAY);
+          cell.setCellStyle(getDateTimeStyle(workbook, "HH:mm:ss"));
+        }
+        default -> {
+          if (ensoToTextCallback != null) {
+            cell.setCellValue(ensoToTextCallback.apply(value));
+          } else {
+            throw new IllegalArgumentException("Enso to text callback is not set. Unable to process value.");
+          }
+        }
+      }
+    }
+  }
+
+  private static boolean hasHeaders(ExcelSheet excelSheet, int topRow, int startCol, int endCol) {
```

Comment:
This raises an interesting design question.

What happens if we load a sheet which had no headers, so our column names are from the default excel headers (`A`, `B`, ...). Then we perform some manipulation on the data, but no column names get changed and _then_ we save the table into a _completely new_ excel file (the file did not exist prior), with `headers=Infer`.

What will happen? Will the `A`, `B`, ... headers be written as a first row? Or will they be ignored?

It seems intuitively that if the headers were taken from default excel headers they should not be written to the new sheet either, but currently we cannot really track this without some special flag in the table. It gets even more complicated if the columns _are_ modified - if some column is removed or a column with a new name gets added - in that case it seems to make sense to actually do write the headers into the file as they won't match the defaults anymore.

I'm not sure what is the right solution here and I'm not saying that it needs to be solved with this PR, but this piece of code has raised this question and I thought it is worth considering at some point.

Reply:
Yes if `Headers=Infer` it would write `A`,`B`,`C`, etc.

I think this is a valid output but yes see in this odd case results in an odd header row being written.

Reply:
Do you think it may make sense to check in Infer+Excel mode if the column names (taking order into account) are exactly like the auto-generated Excel ones and in that case ignoring them? Just thinking out loud, but I guess in practice it may be more confusing to the users and they can always just set it to explicit `False` to have what's needed, so I guess it's not worth it...

Reply:
I think it would be more confusing. Better to make the users choose to explicitly not write headers.